consensus/core/src/core.rs

```diff
@@ -749,6 +812,193 @@ impl Core {
+        self.ancestor_state_manager.update_all_ancestors_state();
+        let ancestor_state_map = self.ancestor_state_manager.get_ancestor_states();
+
+        let quorum_round = clock_round.saturating_sub(1);
+
+        let mut temp_excluded_ancestors = Vec::new();
+
+        // Propose only ancestors of higher rounds than what has already been proposed.
+        // And always include own last proposed block first among ancestors.
+        // Start by only including the high scoring ancestors. Low scoring ancestors
+        // will be included in a second pass below.
+        let included_ancestors = iter::once(self.last_proposed_block.clone())
+            .chain(
+                ancestors
+                    .into_iter()
+                    .filter(|ancestor| ancestor.author() != self.context.own_index)
+                    .flat_map(|ancestor| {
+
+                        let ancestor_state = ancestor_state_map[ancestor.author()];
+
+                        match ancestor_state {
+                            AncestorState::Include => {
+                                debug!("Found ancestor {ancestor} with INCLUDE state for round {clock_round}");
+                            }
+                            AncestorState::Exclude(score) => {
+                                debug!("Added ancestor {ancestor} with EXCLUDE state with score {score} to temporary excluded ancestors for round {clock_round}");
+                                temp_excluded_ancestors.push((score, ancestor));
+                                return None;
+                            }
+                        }
+
+                        if let Some(last_block_ref) =
+                            self.last_included_ancestors[ancestor.author()]
+                        {
+                            return (last_block_ref.round < ancestor.round()).then_some(ancestor);
+                        }
+                        Some(ancestor)
+                    }),
+            )
+            .collect::<Vec<_>>();
+
+        let mut parent_round_quorum = StakeAggregator::<QuorumThreshold>::new();
+
+        // Check total stake of high scoring parent round ancestors
+        for ancestor in included_ancestors
+            .iter()
+            .filter(|a| a.round() == quorum_round)
+        {
+            parent_round_quorum.add(ancestor.author(), &self.context.committee);
+        }
+
+        if smart_select && !parent_round_quorum.reached_threshold(&self.context.committee) {
+            tracing::error!("code_cov-5");
+            self.context.metrics.node_metrics.smart_selection_wait.inc();
+            debug!("Only found {} stake of good ancestors to include for round {clock_round}, will wait for more.", parent_round_quorum.stake());
+            return vec![];
+        }
+
+        // Now we can update the last included ancestor block refs as there is no way
+        // to short circuit ancestor proposal.
+        for ancestor in included_ancestors.iter() {
```

Comment:
Should this logic run at the end, maybe even after the smart_ancestors_to_select() / ancestors_to_select() calls?